File: 2025-12/2025-12-06/date_formatter.py
import re

from datetime import datetime

def format_date(date_string):
    date_string = re.sub("(?P<year>\d{1,4})$", replacer, date_string)
    date = datetime.strptime(date_string, "%B %d, %Y")
    
    return date.strftime("%Y-%m-%d")

def replacer(match):
    return match[0].zfill(4)

print(format_date("December 6, 2025"))
print(format_date("January 1, 2000"))
print(format_date("November 11, 1111"))
print(format_date("September 7, 512"))
print(format_date("May 4, 1950"))
print(format_date("February 29, 1992"))

# --- TEST SUITE ---

tests_text = '''
1. format_date("December 6, 2025") should return "2025-12-06".
2. format_date("January 1, 2000") should return "2000-01-01".
3. format_date("November 11, 1111") should return "1111-11-11".
4. format_date("September 7, 512") should return "512-09-07".
5. format_date("May 4, 1950") should return "1950-05-04".
6. format_date("February 29, 1992") should return "1992-02-29".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+)\.$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 12 December/12-06/dateFormatter.js
function formatDate(dateString) {

  return new Date(dateString).toLocaleDateString("en-CA");
}

console.log(formatDate("December 6, 2025"));
console.log(formatDate("January 1, 2000"));
console.log(formatDate("November 11, 1111"));
console.log(formatDate("September 7, 512"));
console.log(formatDate("May 4, 1950"));
console.log(formatDate("February 29, 1992"));

// --- TEST SUITE ---

const testsText = `
1. formatDate("December 6, 2025") should return "2025-12-06".
2. formatDate("January 1, 2000") should return "2000-01-01".
3. formatDate("November 11, 1111") should return "1111-11-11".
4. formatDate("September 7, 512") should return "512-09-07".
5. formatDate("May 4, 1950") should return "1950-05-04".
6. formatDate("February 29, 1992") should return "1992-02-29".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+)\.$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);